### DISCUSSION TITLE
Work with IQueryables in AutoMapper

### DISCUSSION THREAD

**Discussion by abdollahkahne** [LuckyPennySoftware/AutoMapper#3779] [NA]
Hi,
consider that I have a query which return `IQueryable<Company>`  and I should map it to `IQueryable<CompanyDto>`. When I use the **AutoMapper**, it sends the all-columns query to database and then, after returning the result, it try to apply the requested mapping which is not a good practice considering performance. Does exist any method that support this situation?
To clarify my question I send a similar but simple Example:
```
public class Company
    {
        [Column("CompanyId")]
        public Guid Id { get; set; }
        [Required(ErrorMessage = "Company name is a required field.")]
        [MaxLength(60, ErrorMessage = "Maximum length for the Name is 60 characters.")]
        public string Name { get; set; }

        [Required(ErrorMessage = "Company address is a required field.")]
        [MaxLength(60, ErrorMessage = "Maximum length for the Address is 60 characters.")]
        public string Address { get; set; }
        public string Country { get; set; }
        public ICollection<Employee> Employees { get; set; }
    }
```
and
```

public class CompanyDto
    {
        public string FullAddress { get; set; }
    } 
```
AutoMapper Profile:

```
 public class AutoMapperProfile : Profile
    {
        public AutoMapperProfile()
        {
            CreateMap<Company, CompanyDto>()
            .ForMember(company => company.FullAddress, memberOptions => memberOptions.MapFrom(c => $"{c.Address} {c.Country}"));
        }
    }
```
Now I have two alternative to make the query:
**1- not using AutoMapper as below:**

```
var companies = _repository.Company.GetAll(false);
                var companiesDTO = companies.Select(c => new CompanyDto
                {
                    // Id = c.Id,
                    // Name = c.Name,
                    FullAddress = $"{c.Address} {c.Country}"
                }).AsEnumerable();
```
As the following image shows it only request the required columns:
![image](https://user-images.githubusercontent.com/48277995/141925360-db422e52-cd25-4ca5-a20a-8b6ebb803dd2.png)

**2-Using the AutoMapper**
`var companiesDTO = _mapper.Map<IQueryable<Company>, IEnumerable<CompanyDto>>(companies); // Altough this work but I am not sure that works performant`

and the query that sent to database is the full columns query
![image](https://user-images.githubusercontent.com/48277995/141925610-50711385-d529-4e03-87e7-6bb6ee1f2515.png)

**Comment by lbargaoanu** [comment]
https://docs.automapper.org/en/latest/Queryable-Extensions.html

**Comment by abdollahkahne** [comment]
Yes, after adding `ProjectTo` directly to query, it works as expected. Thanks
just for adding more information to question:
`var companies = _repository.Company.GetAll(false).ProjectTo<CompanyDto>(_mapper.ConfigurationProvider);`

and the resulted executed query:
![image](https://user-images.githubusercontent.com/48277995/141998111-74a7e33c-ca49-492e-925a-0ce5d194910e.png)
